ucb_prefect_tools/database.py

```
@@ -546,6 +538,100 @@ def _oracle_execute_statements(conn, cursor, host, statements, query_params=None
         raise
 
 
+##########
+# Snowflake functions
+##########
+
+
+def snowflake_insert(
+    dataframe: pd.DataFrame,
+    table_identifier: str,
+    connection_info: dict,
+    pre_insert_statements: list[str] = None,
+    pre_insert_params: list = None,
+    max_error_proportion: float = 0.05,  # Not implemented!
```

Do you see a need to implement in this in the future? Or would you say any transaction should be failed completely if errors crop up?

---

So by default Snowflake will throw an exception, stop loading, and revert the transaction if a single error is discovered. I went back and forth and ultimately thought it was best to just go with that default. You can also tell it to skip the file if more than X% of it is bad, but they caution against using that blindly since the entire file will be buffered regardless of the error rate and this could lead to wasted compute. You could also dry-run import the file to get the exact number of errors before doing the real import, but again this seems like a waste of compute. I'm thinking the best approach is to just let it fail on the first error, and then if someone wants more information than the one error message, they can do their own dry-run import.